internal/storage/serde.go

```diff
@@ -0,0 +1,414 @@
+	Record() Record
+	Close()
+}
+
+// compositeRecord is a record being composed of multiple records, in which each only have 1 column
+type compositeRecord struct {
+	recs   map[FieldID]arrow.Record
+	schema map[FieldID]schemapb.DataType
+}
+
+func (r *compositeRecord) Column(i FieldID) arrow.Array {
+	return r.recs[i].Column(0)
+}
+
+func (r *compositeRecord) Len() int {
+	for _, rec := range r.recs {
+		return rec.Column(0).Len()
+	}
+	return 0
+}
+
+func (r *compositeRecord) Release() {
+	for _, rec := range r.recs {
+		rec.Release()
+	}
+}
+
+func (r *compositeRecord) Schema() *map[FieldID]schemapb.DataType {
+	return &r.schema
+}
+
+type compositeRecordReader struct {
+	RecordReader
+	blobs [][]*Blob
+
+	blobPos int
+	rrs     []array.RecordReader
+	closers []func()
+	fields  []FieldID
+
+	r compositeRecord
+}
+
+func (crr *compositeRecordReader) iterateNextBatch() error {
+	if crr.closers != nil {
+		for _, close := range crr.closers {
+			if close != nil {
+				close()
+			}
+		}
+	}
+	crr.blobPos++
+	if crr.blobPos >= len(crr.blobs[0]) {
+		return io.EOF
+	}
+
+	for i, b := range crr.blobs {
+		reader, err := NewBinlogReader(b[crr.blobPos].Value)
+		if err != nil {
+			return err
+		}
+
+		crr.fields[i] = reader.FieldID
+		// TODO: assert schema being the same in every blobs
+		crr.r.schema[reader.FieldID] = reader.PayloadDataType
+		er, err := reader.NextEventReader()
+		if err != nil {
+			return err
+		}
+		rr, err := er.GetArrowRecordReader()
+		if err != nil {
+			return err
+		}
+		crr.rrs[i] = rr
+		crr.closers[i] = func() {
+			rr.Release()
+			er.Close()
+			reader.Close()
+		}
+	}
+	return nil
+}
+
+func (crr *compositeRecordReader) Next() error {
+	if crr.rrs == nil {
+		if crr.blobs == nil || len(crr.blobs) == 0 {
+			return io.EOF
+		}
+		crr.rrs = make([]array.RecordReader, len(crr.blobs))
+		crr.closers = make([]func(), len(crr.blobs))
+		crr.blobPos = -1
+		crr.fields = make([]FieldID, len(crr.rrs))
+		crr.r = compositeRecord{
+			recs:   make(map[FieldID]arrow.Record, len(crr.rrs)),
+			schema: make(map[FieldID]schemapb.DataType, len(crr.rrs)),
+		}
+		crr.iterateNextBatch()
+	}
+
+	composeRecord := func() bool {
+		for i, rr := range crr.rrs {
+			if ok := rr.Next(); !ok {
+				return false
+			}
+			// compose record
+			crr.r.recs[crr.fields[i]] = rr.Record()
+		}
+		return true
+	}
+
+	// Try compose records
+	if ok := composeRecord(); !ok {
+		// If failed the first time, try iterate next batch (blob), the error may be io.EOF
+		if err := crr.iterateNextBatch(); err != nil {
+			return err
+		}
+		// If iterate next batch success, try compose again
+		if ok := composeRecord(); !ok {
+			// If the next blob is empty, return io.EOF (it's rare).
+			return io.EOF
+		}
+	}
+	return nil
+}
+
+func (crr *compositeRecordReader) Record() Record {
+	return &crr.r
+}
+
+func (crr *compositeRecordReader) Close() {
+	for _, close := range crr.closers {
+		close()
+	}
+}
+
+func getBlobColLog(blob *Blob) (FieldID, UniqueID) {
```

Comment:
maybe we should return an error if the length of splitted blob.Key is less than 2?